dm/dm/master/scheduler/scheduler_test.go

```diff
@@ -1353,6 +1353,82 @@ func (t *testScheduler) TestStartStopRelay(c *C) {
 	c.Assert(bound, IsFalse)
 }
 
+func (t *testScheduler) TestRelayWithWithoutWorker(c *C) {
+	defer clearTestInfoOperation(c)
+
+	var (
+		logger      = log.L()
+		s           = NewScheduler(&logger, config.Security{})
+		sourceID1   = "mysql-replica-1"
+		workerName1 = "dm-worker-1"
+		workerName2 = "dm-worker-2"
+	)
+
+	worker1 := &Worker{baseInfo: ha.WorkerInfo{Name: workerName1}}
+	worker2 := &Worker{baseInfo: ha.WorkerInfo{Name: workerName2}}
+
+	// step 1: start an empty scheduler
+	s.started = true
+	s.etcdCli = etcdTestCli
+	s.workers[workerName1] = worker1
+	s.workers[workerName2] = worker2
+	s.sourceCfgs[sourceID1] = &config.SourceConfig{}
+
+	worker1.ToFree()
+	c.Assert(s.boundSourceToWorker(sourceID1, worker1), IsNil)
+	worker2.ToFree()
+
+	// step 2: check when enable-relay = false, can start/stop relay without worker name
+	c.Assert(s.StartRelay(sourceID1, []string{}), IsNil)
+	c.Assert(s.sourceCfgs[sourceID1].EnableRelay, IsTrue)
+
+	c.Assert(s.StartRelay(sourceID1, []string{}), IsNil)
+	c.Assert(s.sourceCfgs[sourceID1].EnableRelay, IsTrue)
+
+	c.Assert(s.StopRelay(sourceID1, []string{}), IsNil)
+	c.Assert(s.sourceCfgs[sourceID1].EnableRelay, IsFalse)
+
+	c.Assert(s.StopRelay(sourceID1, []string{}), IsNil)
+	c.Assert(s.sourceCfgs[sourceID1].EnableRelay, IsFalse)
+
+	// step 3: check when enable-relay = false, can start/stop relay with worker name
+	c.Assert(s.StartRelay(sourceID1, []string{workerName1, workerName2}), IsNil)
+	c.Assert(s.sourceCfgs[sourceID1].EnableRelay, IsFalse)
```

Comment:
why sourcecfg1.EnableRelay is False 🤔 (do we need also update EnableRelay filed when start/stop-relay with worker name)

Reply:
by design